scp/cc/cpio/client_providers/private_key_fetcher_provider/src/azure/azure_private_key_fetcher_provider_utils.cc

```diff
@@ -29,16 +29,391 @@ using google::scp::core::Uri;
+
+ private:
+  RSA* rsa_;
+};
+
+class BnWrapper {
+ public:
+  BnWrapper() : bn_(BN_new()) {}
+
+  ~BnWrapper() { BN_free(bn_); }
+
+  BIGNUM* get() { return bn_; }
+
+ private:
+  BIGNUM* bn_;
+};
+
+class EvpPkeyWrapper {
+ public:
+  EvpPkeyWrapper() : pkey_(EVP_PKEY_new()) {}
+
+  ~EvpPkeyWrapper() { EVP_PKEY_free(pkey_); }
+
+  EVP_PKEY* get() { return pkey_; }
+
+ private:
+  EVP_PKEY* pkey_;
+};
+
+class BIOWrapper {
+ public:
+  explicit BIOWrapper(BIO_METHOD* method) : bio_(BIO_new(method)) {}
+
+  ~BIOWrapper() { BIO_free(bio_); }
+
+  BIO* get() { return bio_; }
+
+ private:
+  BIO* bio_;
+};
+
+class EVPKeyCtxWrapper {
+public:
+    explicit EVPKeyCtxWrapper(EVP_PKEY_CTX* ctx) : ctx_(ctx) {}
+    ~EVPKeyCtxWrapper() {
+        if (ctx_) {
+            EVP_PKEY_CTX_free(ctx_);
+            ctx_ = nullptr;
+        }
+    }
+    
+    EVP_PKEY_CTX* get() const { return ctx_; }
+
+private:
+    EVP_PKEY_CTX* ctx_;
+};
+
+bool AzurePrivateKeyFetchingClientUtils::isPrivate(EVP_PKEY* pkey) {
+  // Determine if the key is private or public
+  int key_type = EVP_PKEY_type(pkey->type);
+  bool is_private = false;
+  if (key_type == EVP_PKEY_RSA) {
+    const RSA* rsa_key = EVP_PKEY_get1_RSA(pkey);
+    is_private =
+        (rsa_key->e != NULL && rsa_key->n != NULL && rsa_key->d != NULL);
+  }
+  return is_private;
+}
+
 void AzurePrivateKeyFetchingClientUtils::CreateHttpRequest(
     const PrivateKeyFetchingRequest& request, HttpRequest& http_request) {
   const auto& base_uri =
       request.key_vending_endpoint->private_key_vending_service_endpoint;
   http_request.method = HttpMethod::POST;
 
   http_request.path = std::make_shared<Uri>(base_uri);
+
+  // Generate attestation report
   const auto report =
       hasSnp() ? fetchSnpAttestation() : fetchFakeSnpAttestation();
   CHECK(report.has_value()) << "Failed to get attestation report";
-  http_request.body = core::BytesBuffer(nlohmann::json(report.value()).dump());
+
+  nlohmann::json json_obj;
+  json_obj["attestation"] = report.value();
+
+  http_request.body = core::BytesBuffer(json_obj.dump());
 }
+
+/**
+ * @brief Generate a new wrapping key
+ */
+std::pair<EVP_PKEY*, EVP_PKEY*>
+AzurePrivateKeyFetchingClientUtils::GenerateWrappingKey() {
+  RsaWrapper rsa;
+  BnWrapper e;
+
+  BN_set_word(e.get(), RSA_F4);
+  RSA_generate_key_ex(rsa.get(), 4096, e.get(), NULL);
+
+  EvpPkeyWrapper private_key;
+  if (EVP_PKEY_set1_RSA(private_key.get(), rsa.get()) != 1) {
+    char* error_string = ERR_error_string(ERR_get_error(), NULL);
+    throw std::runtime_error(std::string("Getting private EVP_PKEY failed: ") +
+                             error_string);
+  }
+
+  // Create a new EVP_PKEY for the public key
+  EvpPkeyWrapper public_key;
+  const RSA* rsa_pub = EVP_PKEY_get1_RSA(private_key.get());
+  RSA* rsa_pub_dup = RSA_new();
+  if (!RSA_set0_key(rsa_pub_dup, rsa_pub->n, rsa_pub->e, NULL)) {
+    char* error_string = ERR_error_string(ERR_get_error(), NULL);
```

Comment:
When you get openssl error, I think it's safer to call `ERR_clear_error` before calling functions that can return errors.  just in case other code forgets to clear it. Missing call for this can lead to difficult debugging time like you get an error message "key type is wrong" while the actual error is "key value is empty".
Also I think you want to call it when you get an error.

Reply:
added ERR_clear_error